### PR TITLE
encapsulate the agent mapping operations

### PR DESCRIPTION
**Pull request by chriselion**
Trying to clean up and encapsulate these operations. If the initial approach sounds good, I think I can make all the operations O(1).

### CODE REVIEW

File: gym-unity/gym_unity/tests/test_gym.py [hunk 1]
```diff
@@ -183,3 +188,33 @@ def setup_mock_unityenvironment(mock_env, mock_spec, mock_result):
     mock_env.return_value.get_agent_groups.return_value = ["MockBrain"]
     mock_env.return_value.get_agent_group_spec.return_value = mock_spec
     mock_env.return_value.get_step_result.return_value = mock_result
+
+
+@pytest.mark.parametrize("mapper_cls", [AgentIdIndexMapper, AgentIdIndexMapperSlow])
```
Comment (vincentpierre) [inline gym-unity/gym_unity/tests/test_gym.py]: That's pretty cool.

Comment (chriselion) [reply gym-unity/gym_unity/tests/test_gym.py]: Do you mean `pytest.mark.parametrize`, or passing class types as parameters?

File: gym-unity/gym_unity/tests/test_gym.py [hunk 2]
```diff
@@ -183,3 +188,33 @@ def setup_mock_unityenvironment(mock_env, mock_spec, mock_result):
...
+
+    # Order of the rewards don't matter
+    assert {old_reward1, old_reward2} == {42.0, 1337.0}
+
+    new_agent_ids = [1002, 1003, 2001, 2002]
+    permutation = mapper.get_id_permutation(new_agent_ids)
+    # Make sure it's actually a permutation - needs to contain 0..N-1 with no repeats.
+    assert set(permutation) == set(range(0, 4))
+
+    # For initial agents that were in the initial group, they need to be in the same slot.
```
Comment (chriselion) [inline gym-unity/gym_unity/tests/test_gym.py]: @vincentpierre Is my claim here correct about the new agents? Order shouldn't matter as long as it's consistent, right?

File: gym-unity/gym_unity/tests/test_gym.py (hunk 1 above)
Comment (vincentpierre) [reply gym-unity/gym_unity/tests/test_gym.py]: both

File: gym-unity/gym_unity/tests/test_gym.py (hunk 2 above)
Comment (vincentpierre) [reply gym-unity/gym_unity/tests/test_gym.py]: Yes, although anywhere might be confusing. They must replace previously done agents

File: gym-unity/gym_unity/tests/test_gym.py [hunk 3]
```diff
@@ -183,3 +188,33 @@ def setup_mock_unityenvironment(mock_env, mock_spec, mock_result):
...
+    # Agents that were added later can be anywhere.
```
Comment (chriselion) [inline gym-unity/gym_unity/tests/test_gym.py]:
```suggestion
    # Agents that were added later can be be in any free slot (from a previous agent).
```

**Pull request merged**